pennylane/devices/experimental/device_api.py

```diff
@@ -0,0 +1,555 @@
+
+
+# pylint: disable=unused-argument, no-self-use
+class Device(abc.ABC):
+    """A device driver that can control one or more backends. A backend can be either a physical
+    Quantum Processing Unit or a virtual one such as a simulator.
+
+    .. warning::
+
+        This interface is **experimental** and not yet integrated with the rest of PennyLane.
+
+    Device drivers should be configured to run under :func:`~.enable_return`, the newer
+    return shape specification.
+
+    Only the ``execute`` method must be defined to construct a device driver.
+
+    .. details::
+        :title: Design Motivation
+
+        **Streamlined interface:** Only methods that are required to interact with the rest of PennyLane will be placed in the
+        interface. Developers will be able to clearly see what they can change while still having a fully functional device.
+
+        **Reduction of duplicate methods:** Methods that solve similar problems are combined together. Only one place will have
+        to solve each individual problem.
+
+        **Support for dynamic execution configurations:** Properties such as shots belong to specific executions.
+
+        **Greater coverage for differentiation methods:** Devices can define any order of derivative, the vector jacobian product,
+        or the jacobian vector product.  Calculation of derivatives can be done at the same time as execution to allow reuse of intermediate
+        results.
+
+    .. details::
+        :title: Porting from the old interface
+
+        :meth:`~.Device.batch_execute` and :meth:`~.Device.execute` are now a single method, :meth:`~.Device.execute`
+
+        :meth:`~.Device.batch_transform` and :meth:`~.Device.expand_fn` are now a single method, :meth:`~.Device.preprocess`
+
+        Shot information is no longer stored on the device, but instead specified on individual input :class:`~.QuantumTape`.
+
+        The old devices defined a :meth:`~.Device.capabilities` dictionary that defined characteristics of the devices and controlled various
+        preprocessing and validation steps, such as ``"supports_broadcasting"``.  These capabilites should now be handled by the
+        :meth:`~.Device.preprocess` method. For example, if a device does not support broadcasting, ``preprocess`` should
+        split a quantum script with broadcasted parameters into a batch of quantum scripts. If the device does not support mid circuit
+        measurements, then ``preprocess`` should apply :func:`~.defer_measurements`.  A set of default preprocessing steps will be available
+        to make a seemless transition to the new interface.
+
+        A class will be provided to easily construct default preprocessing steps from supported operations, supported observables,
+        supported measurement processes, and various capabilities.
+
+        Utility functions will be added to the ``devices`` module to query whether or not the device driver can do certain things, such
```

Comment:
"devices module" might be ambiguous while this module remains experimental.

Reply:
We can go back and clarify a lot of these details when they actually exist somewhere.